File: VFIT_Backend/src/main/java/com/vfit/modules/exercise_library/document/ExerciseCatalog.java
package com.vfit.modules.exercise_library.document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "exercise_catalogs")
@CompoundIndexes({
        @CompoundIndex(name = "idx_version_active", def = "{ 'version': -1, 'active': 1 }")
})
public class ExerciseCatalog {
    @Id
    private String id;
    private int version;
    @Indexed
    private String locale;
    @Indexed
    private boolean active;
    @Builder.Default
    private List<MuscleGroup> groups = new ArrayList<>();
    @CreatedDate
    private Instant createdAt;
    @LastModifiedDate
    private Instant updatedAt;

    @Getter
    @Setter
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class MuscleGroup {
        private String slug;
        private String name;
        private int sortOrder;
        @Builder.Default
        private List<SubMuscleGroup> subGroups = new ArrayList<>();
    }

    @Getter
    @Setter
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class SubMuscleGroup {
        private String slug;
        private String name;
        private int sortOrder;
        @Builder.Default
        private List<ExerciseItem> exercises = new ArrayList<>();
    }

    @Getter
    @Setter
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ExerciseItem {
        private String id;
        private String name;
        private String videoUrl;
        private String videoSearchHint;
        private String description;
        @Builder.Default
        private List<String> equipment = new ArrayList<>();
        private int sortOrder;
        @Builder.Default
        private boolean active = true;
    }
}

File: VFIT_Fontend/lib/features/exercise_library/domain/entities/exercise_catalog.dart
import 'package:equatable/equatable.dart';

class ExerciseCatalog extends Equatable {
  const ExerciseCatalog({
    required this.catalogVersion,
    required this.groups,
    this.isStale = false,
  });

  final int catalogVersion;
  final List<MuscleGroup> groups;
  final bool isStale;

  ExerciseCatalog copyWith({bool? isStale}) {
    return ExerciseCatalog(
      catalogVersion: catalogVersion,
      groups: groups,
      isStale: isStale ?? this.isStale,
    );
  }

  @override
  List<Object?> get props => [catalogVersion, groups, isStale];
}

class MuscleGroup extends Equatable {
  const MuscleGroup({
    required this.id,
    required this.name,
    required this.subGroups,
  });

  final String id;
  final String name;
  final List<SubMuscleGroup> subGroups;

  int get exerciseCount => subGroups.fold<int>(
        0,
        (total, subGroup) => total + subGroup.exercises.length,
      );

  @override
  List<Object?> get props => [id, name, subGroups];
}

class SubMuscleGroup extends Equatable {
  const SubMuscleGroup({
    required this.id,
    required this.name,
    required this.exercises,
  });

  final String id;
  final String name;
  final List<ExerciseItem> exercises;

  @override
  List<Object?> get props => [id, name, exercises];
}

class ExerciseItem extends Equatable {
  const ExerciseItem({
    required this.id,
    required this.name,
    required this.equipment,
    this.videoUrl,
    this.videoSearchHint,
    this.description,
  });

  final String id;
  final String name;
  final String? videoUrl;
  final String? videoSearchHint;
  final String? description;
  final List<String> equipment;

  @override
  List<Object?> get props => [
        id,
        name,
        videoUrl,
        videoSearchHint,
        description,
        equipment,
      ];
}
